### PR TITLE
TimeKeeper alternate integration

### PR DESCRIPTION
**Pull request by joelwetzel**
This is an alternative to: https://github.com/biocomp/hubitat_ci/pull/52

Instead of overriding the default constructor of Date(), this replaces "new Date()" during parsing with "new TimeKeeperDate()"

### CODE REVIEW

File: src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        given:
+        def dimmerFixture = DimmerFixtureFactory.create('n')
+
+        // Run the app sandbox, passing the dimmer fixture in.
+        def appScript = sandbox.run(api: appExecutor,
+            userSettingValues: [dimmers: [dimmerFixture], minimumLevel: 5, enableLogging: true])
+        appExecutor.setSubscribingScript(appScript)
+
+        when:
+        def now = new Date()
```
Comment (biocomp) [inline src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy:39]: The date can become different between calling `now()` and `scriptNow()`, the test is flaky.
I guess a more reliable test would be to set the TimeKeeper to some date in the past from `now()` and check that `scriptNow()` returns a different date.

File: src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        def timekeeper = new TimeKeeper(Date.parse("yyyy-MM-dd hh:mm:ss", "2014-08-31 8:23:45"))
```
Comment (biocomp) [inline src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy:57]: Can install() and uninstall() be done in constructor/"destructor" of TimeKeeper? It seems like it's pretty easy to forget to uninstall it.

I think another approach would make the `TimeKeeperDate` read the date from some global static variable always, then if your test wants to change it, it'd have to change that static variable.

Yet another approach could be making `TimeKeeperDate` take the script object in its constructor, and then pass the fake date through the script object, so that the global variable is not involved. But then there would be issues with mocking static methods...

But since we are generating the script class on the fly, we could attach the static variable to the script class... maybe?

Well, anyway, something to think about.

File: src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy [hunk 3]
```diff
@@ -36,11 +36,13 @@ class AppValidator extends ValidatorBase{
...
 
     HubitatAppScript parseScript(File scriptFile) {
-        return constructParser(HubitatAppScript).parse(scriptFile) as HubitatAppScript
+        def scriptFileText = scriptFile.getText('UTF-8')
+        return parseScript(scriptFileText)
     }
 
     HubitatAppScript parseScript(String scriptText) {
-        return constructParser(HubitatAppScript).parse(scriptText) as HubitatAppScript
+        scriptText = scriptText.replaceAll("new\\s*Date\\s*\\(\\s*\\)", "new me.biocomp.hubitat_ci.util.TimeKeeperDate()")
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy:44]: Move this out into a separate method (like `fixupScriptText` in `ValidatorBase`), and reuse it between `AppValidator` and `DeviceValidator`. We need the same mocking functionality for devices too, right? And add a simple test to see that it works for device scripts.

File: src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy (hunk 1 above)
Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy:39]: You are correct, there can be a couple milliseconds in there.  I'll adjust the test to allow that.

File: src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy (hunk 2 above)
Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy:58]: So... groovy doesn't have a destructor.  We do need the install/uninstall, unfortunately.  Install intercepts the metaclass of TimeKeeperDate, and uninstall restores it.  And that must be done because we only want this substitution to happen during tests that we turn it on.  (And since Spock runs all the tests in a single runtime, if a single test installs it but doesn't uninstall it, that can have effects on other tests that follow.)

One of my early attempts at this (even before I found your framework) was to try to intercept the "new Date()" inside of an app script.  However, it's a huge minefield, and required the app script author to understand a lot about how time updates were happening.  This design is to avoid that.  The idea being that I can take any off-the-shelf app script that depends on time, and write tests for it immediately.

(You'll see more of that in an upcoming PR.  I'm implementing the BaseScheduler portion of AppExecutor.)

File: src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy (hunk 3 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy]: Done.

File: src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy [hunk 4]
```diff
@@ -36,11 +36,13 @@ class AppValidator extends ValidatorBase{
...
 
     HubitatAppScript parseScript(File scriptFile) {
-        return constructParser(HubitatAppScript).parse(scriptFile) as HubitatAppScript
+        def scriptFileText = scriptFile.getText('UTF-8')
+        return parseScript(scriptFileText)
     }
 
     HubitatAppScript parseScript(String scriptText) {
-        return constructParser(HubitatAppScript).parse(scriptText) as HubitatAppScript
+        scriptText = super.replaceDateReferences(scriptText)
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy:44]: I'd call this something more generic for any potential future script patches.
```suggestion
        scriptText = super.patchScriptText(scriptText)
```

File: src/main/groovy/me/biocomp/hubitat_ci/validation/ValidatorBase.groovy [hunk 5]
```diff
@@ -293,4 +295,9 @@ class ValidatorBase {
     boolean hasAnyOfFlags(Set<Flags> flags) {
         return this.flags.intersect(flags)
     }
+
+    String replaceDateReferences(String scriptText) {
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/validation/ValidatorBase.groovy:299]:
```suggestion
    static String replaceDateReferences(String scriptText) {
```

File: src/test/groovy/me/biocomp/hubitat_ci/integration/AppDateSubstitutionTest.groovy [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            userSettingValues: [dimmers: [dimmerFixture], minimumLevel: 5, enableLogging: true])
+        appExecutor.setSubscribingScript(appScript)
+
+        when:
+        def now = new Date()
+        def nowAccordingToTheAppScript = appScript.scriptNow()
+        def differenceInMillis = nowAccordingToTheAppScript.getTime() - now.getTime()
+
+        then: "Since we're generating Dates twice in succession, there can be a slight time difference, but we don't want to allow much."
+        differenceInMillis < 10
```
Comment (biocomp) [inline src/test/groovy/me/biocomp/hubitat_ci/integration/AppDateSubstitutionTest.groovy:44]: I'd give it much larger diff. Maybe up to a second. Who knows how slow the servers can be.

File: src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy (hunk 2 above)
Comment (biocomp) [reply src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy]: I think things can be simplified at least like so:

```
class TimeKeeper{
  public static Date current_date = new Date()
}

// Somewhere in the test:
TimeKeeper.current_date = Date.parse("yyyy-MM-dd hh:mm:ss", "2014-08-31 8:23:45")

// In scripts, replace new Date() with:
println TimeKeeper.current_date
```

Comment (biocomp) [reply src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy]: Probably needs to be made thread-safe

File: src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy (hunk 4 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy]: Done

File: src/main/groovy/me/biocomp/hubitat_ci/validation/ValidatorBase.groovy (hunk 5 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/validation/ValidatorBase.groovy]: Done

File: src/test/groovy/me/biocomp/hubitat_ci/integration/AppDateSubstitutionTest.groovy (hunk 6 above)
Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/integration/AppDateSubstitutionTest.groovy]: Done

File: src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy (hunk 2 above)
Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/integration/DateSubstitutionTest.groovy]: Ok, I just pushed up a commit that TimeKeeper into a thread-safe static class.  This removes the need for the TimeKeeperDate class, timekeeper.install(), and timekeeper.uninstall().  Also added in TimeKeeper observability, which will be utilized by upcoming PRs.

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    TimeKeeper() {
+        throw new Exception("TimeKeeper is a static class and should not be instantiated.")
+    }
+
+    static Date now() {
+        return internalDate
+    }
+
+    static void set(Date newDate) {
+        internalDate = newDate
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:29]: Seems like accessing the `internalDate` needs to be thread safe too. I don't think java guarantees atomic accesses to the references?

File: src/test/groovy/me/biocomp/hubitat_ci/ApiCompatibilityTest.groovy [hunk 8]
```diff
@@ -235,7 +235,7 @@ class ApiCompatibilityTest extends Specification
 
         then:
             result.firmware_version == "1.2.3"
-            !classes.find { it.class_name == "hubitat_api_exporter_app" }.methods.isEmpty() //!= startsWith('[{"class_name":"hubitat_api_exporter_app"')//"[[class_name:'hubitat_api_exporter_app',methods:[]]]"
```
Comment (biocomp) [inline src/test/groovy/me/biocomp/hubitat_ci/ApiCompatibilityTest.groovy:238]: Removing of commented out code is fine, but please keep the `hubitat_api_exporter_app` there.
BTW, why did you need to add "Script1" to the list of exceptions?

Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/ApiCompatibilityTest.groovy:238]: This is a result of needing to find-and-replace in the script text.  Previously, if you loaded a script using the File class, the GroovyShell would name the new class using the filename, which is why it was hubitat_api_exporter_app before.  Maybe I can make it still read the filename from the file, so that you can keep the test the same.

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy (hunk 7 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:29]: You are correct.  Adding this now.

File: src/test/groovy/me/biocomp/hubitat_ci/ApiCompatibilityTest.groovy (hunk 8 above)
Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/ApiCompatibilityTest.groovy:238]: Ok, I was able to get the old behavior back and restore the tests.

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        timeChangedListeners.remove(listener)
+    }
+
+    @Synchronized("timekeeperLock")
+    static void removeAllListeners() {
+        timeChangedListeners.clear()
+    }
+
+    @Synchronized("timekeeperLock")
+    static void fireTimeChangedEvent(Date oldTime, Date newTime) {
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:85]: Should you just do this in setter for the date (make it a property)?

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:85]: I didn't want to fire one if you manually set the date.  That's a sort of "reset".  And to fire a time changed event, you need to know both the previous time and the new time, and they need to be in ascending order.  (That'll be important in the next PR.)

Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:85]: Actually sorry, but can we not have these tracing events in this PR at all? They are kind of unrelated to the current change? And then we can discuss everything in the next more relevant PR?

File: src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy [hunk 10]
```diff
@@ -36,11 +36,19 @@ class AppValidator extends ValidatorBase{
     }
 
     HubitatAppScript parseScript(File scriptFile) {
-        return constructParser(HubitatAppScript).parse(scriptFile) as HubitatAppScript
+        def scriptFileText = scriptFile.getText('UTF-8')
+
+        def scriptName = scriptFile.name
+        if (scriptName.endsWith(".groovy")) {
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy:42]: Should this be more generic to allow for any extension?
Do we have `org.apache.commons.io.FilenameUtils.getBaseName` available (or does it require an extra dependency)?

Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/app/AppValidator.groovy:42]: Also seems like getting script name from filename should also be moved out to a separate function and reused.

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy (hunk 9 above)
Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:85]: Sorry, did you see the last comment here?

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy]: Just saw it this morning.  I've now pushed a commit that removes the observability from TimeKeeper.

File: src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+class TimeKeeperTests extends Specification {
+    def currentDate = new Date()
+
+    def setup() {
+        TimeZone.setDefault(TimeZone.getTimeZone('UTC'))
+
+        TimeKeeper.set(Date.parse("yyyy-MM-dd hh:mm:ss", "2014-08-31 8:23:45"))
+    }
+
+    def cleanup() {
```
Comment (biocomp) [inline src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy:18]: Cleanup not needed?

File: src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+            date2.toString() != currentDate.toString()
+            date2.toString() == "Sun Aug 31 13:23:45 UTC 2014"
+    }
+
+    void "Can advance the days of the TimeKeeper"() {
+        when:
+            def date = TimeKeeper.now()
+
+        then:
+            date.toString() != currentDate.toString()
```
Comment (biocomp) [inline src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy:94]: I think this check is redundant?

Comment (biocomp) [reply src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy:94]: Same in other tests.

**Review by biocomp — APPROVED** [review]

File: src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy (hunk 12 above)
Comment (joelwetzel) [reply src/test/groovy/me/biocomp/hubitat_ci/util/TimeKeeperTests.groovy]: Removed.
